### DISCUSSION TITLE
More flexible config options

### DISCUSSION THREAD

**Discussion by lonix1** [spectreconsole/spectre.console#936] [Ideas]
## Is your feature request related to a problem? Please describe.
The docs show how to create command:
````cs
public class HelloCommand : Command<HelloCommand.Settings>
{
    public class Settings : CommandSettings
    {
        [CommandArgument(0, "[Name]")]               // <------------ here is some config
        public string Name { get; set; }
    }
    // ...
}
````
And how to add it to the app in `Program.Main`:
````cs
var app = new CommandApp();
app.Configure(config =>
{
    config.AddCommand<HelloCommand>("hello")        // <------------ here is more config
        .WithAlias("hola")                          // <------------ etc.
        .WithDescription("Say hello")
        .WithExample(new []{"hello", "Phil"})
        .WithExample(new []{"hello", "Phil", "--count", "4"});
});
````

It's not ideal that the config is separate from the command.

## Describe the solution you'd like
It would be nice to be able to define all config in one place, in the command. For example:

````cs
[CommandName("hello")]
[CommandAlias("holla")]
[CommandDescription("Say hello")]
//...etc
public class HelloCommand : Command<HelloCommand.Settings>
````

That would be good in 80% of the cases. But using attributes means we must use compile-time constants. So it would be nice to also have config in the constructor:
````cs
public HelloCommand() {
  base.Name = "hello";
  base.Alias = "hola";
  base.Description = "hello";
  // etc.
}
````

But there are more case (I've already had one) where you want to call a virtual method, but can't from a ctor, so it would be nice to have config in a virtual method or property:
````cs
public override string GetName() => "hello";

// or some kind of property, maybe something like this:
[CommandName]
public string Name => "hello";
````

Not all the above are necessary, I'm just highlighting the different issues. But it would be beneficial to have the flexibility to define the config where it makes sense for each person:
- in lambdas in the app (easiest, as is done now)
- in attributes on the command (cleaner, but only compile-time constants)
- in the command's ctor (can use non-constants, but cannot easily reuse code via subclassing)
- in the command's config methods or properties (least common, but flexible when reusing code via subclassing)

## Describe alternatives you've considered
Using what is available right now. It's a little messy but it works 100%.

## Additional context
Thank you for this incredible tool, it makes writing a console programme really easy.

**Comment by patriksvensson** [comment]
This is actually by design. The philosophy behind Spectre.Console.Cli is to separate the commands from the settings and tie everything together via composition.

I suspect something else would be more suitable for your needs, such as the excellent [System.CommandLine](https://docs.microsoft.com/en-us/dotnet/standard/commandline/get-started-tutorial). I use it myself when my needs go beyond what Spectre.Console.Cli was designed for.

**Reply by lonix1** [reply]
Very nice of you to tell me about the "competition" :) I never heard of it until now.

It's still a beta though, so I'm going to carry on using your library - which is just great.

**Reply by lonix1** [reply]
BTW I have a workaround for anyone who wants it.

`HelloCommand.cs`
````cs
public class HelloCommand : Command<HelloCommand.Settings>
{
  // ...
  public static string Name => "hello";
  public static string Description => "hello";
  public static string[] Example1 => new string[] { "hello" };
  public static string[] Example2 => new string[] { "hello --foo bar" };
}
````

`Program.cs`
````cs
app.Configure(config => {
  config.AddComand<HelloCommand>(HelloCommand.Name)
    .WithDescription(HelloCommand.Description)
    .WithExample(HelloCommand.Example1)
    .WithExample(HelloCommand.Example2)
    .WithExample(HelloCommand.Example3);                  // <----- not ideal, but works
});
````
That `WithExample` syntax is weird, but not too bad.

**Reply by lonix1** [reply]
One more quick thing: I dismissed System.CommandLine - but then I saw that even though it's a beta, it's been around for two years and has 12 million nuget downloads... incredible.

You said you use it for complex cli stuff - do you feel it's appropriate for non-personal projects (i.e. work), or would you wait for the release version?

**Reply by patriksvensson** [reply]
> You said you use it for complex cli stuff - do you feel it's appropriate for non-personal projects (i.e. work), or would you wait for the release version?

Absolutely. It's stable and works well. Not sure if there are going to be any big API changes before the release, but @jonsequitur probably knows more about that.

**Reply by KathleenDollard** [reply]
The reason we are still in beta is that we are not closed to API changes. Is that enough info? We are currently working on determining what changes we will make, so it's hard for me to be more specific. If your cli is a small part of your app, and are ok making changes, then you can give feedback on System.CommandLine. If you want your app done and do not want to update your code (you user's experience should not change), then perhaps not. 

Also, part of our efforts right now is writing a simplifying layer, so we expect System.CommandLine to be easier to use in the future.
